### PR TITLE
Adds throttling (rate limiting) to the Tahoe API

### PR DESCRIPTION
**Pull request by johnbaldwin**
- User based throttling added to the user registration api
- Throttling is assigned per ViewSet
- Throttle class is `TahoeAPIUserThrottle`
- Rate is 60/minute

### CODE REVIEW

**Comment by johnbaldwin** [comment]
[Add throttling to Tahoe API](https://trello.com/c/HYHlMqtf/677-add-throttling-to-tahoe-api)

File: openedx/core/djangoapps/appsembler/api/permissions.py [hunk 1]
```diff
@@ -44,3 +45,13 @@ class IsSiteAdminUser(BasePermission):
     """
     def has_permission(self, request, view):
         return is_site_admin_user(request)
+
+
+#
+# Throttling
```
Comment (OmarIthawi) [inline openedx/core/djangoapps/appsembler/api/permissions.py]: I'm not a big fan for such commends. I think class doc strings are enough. Would you mind leaving them out?

File: openedx/core/djangoapps/appsembler/api/permissions.py [hunk 2]
```diff
@@ -3,10 +3,11 @@
 Code adapted from Figures
 
 """
-from rest_framework.permissions import BasePermission
-
 import django.contrib.sites.shortcuts
 
+from rest_framework.permissions import BasePermission
+from rest_framework.throttling import UserRateThrottle
```
Comment (OmarIthawi) [inline openedx/core/djangoapps/appsembler/api/permissions.py]: Looks like `UserRateThrottle` already matches what we need, and better than the `RateLimit` module.

File: openedx/core/djangoapps/appsembler/api/tests/test_permissions.py [hunk 3]
```diff
@@ -44,7 +49,8 @@ def get_api_classes():
 
 @ddt.ddt
 class AppsemblerAPIPermissionsTests(TestCase):
-
+    """See the comments in the ``get_api_classes`` function above
```
Comment (OmarIthawi) [inline openedx/core/djangoapps/appsembler/api/tests/test_permissions.py]:
```suggestion
    """
    See the comments in the ``get_api_classes`` function above
```

File: openedx/core/djangoapps/appsembler/api/tests/test_throttling.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
+"""Tests throttling (rate limiting) for the Tahoe API
+
```
Comment (OmarIthawi) [inline openedx/core/djangoapps/appsembler/api/tests/test_throttling.py]:
```suggestion
```

File: openedx/core/djangoapps/appsembler/api/tests/test_throttling.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
+"""Tests throttling (rate limiting) for the Tahoe API
```
Comment (OmarIthawi) [inline openedx/core/djangoapps/appsembler/api/tests/test_throttling.py]:
```suggestion
"""
Tests throttling (rate limiting) for the Tahoe API
```

File: openedx/core/djangoapps/appsembler/api/tests/test_throttling.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    def test_throttle_with_registration_api(self):
+
+        for attempt in xrange(self.rate_limit):
+            post_data = make_post_data(attempt)
+            response = self.client.post(self.url, post_data)
+            self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+        post_data = make_post_data(self.rate_limit)
+        response = self.client.post(self.url, post_data)
+        self.assertEqual(response.status_code, status.HTTP_429_TOO_MANY_REQUESTS)
```
Comment (OmarIthawi) [inline openedx/core/djangoapps/appsembler/api/tests/test_throttling.py]: Looks great, thanks @johnbaldwin!

File: openedx/core/djangoapps/appsembler/api/v1/views.py [hunk 7]
```diff
@@ -77,6 +79,7 @@ class TahoeAuthMixin(object):
 
 class RegistrationViewSet(TahoeAuthMixin, viewsets.ViewSet):
```
Comment (OmarIthawi) [inline openedx/core/djangoapps/appsembler/api/v1/views.py]: Can we get something like this in the API class docstring?

```suggestion
"""
Registration tada yada...

This API endpoint is throttled at the rate of 60 requests per minute. Reaching the limit will cause the API to issue HTTP code `429` (Too many requests). Please allow your API client to handle that case, and be mindful of keep the rate below 60 per minute.

Ta da ya da.
"""
```

File: openedx/core/djangoapps/appsembler/api/permissions.py (hunk 1 above)
Comment (johnbaldwin) [reply openedx/core/djangoapps/appsembler/api/permissions.py]: @OmarIthawi I can remove this

File: openedx/core/djangoapps/appsembler/api/permissions.py (hunk 2 above)
Comment (johnbaldwin) [reply openedx/core/djangoapps/appsembler/api/permissions.py]: @OmarIthawi Then we need to add throttling to global settings. Initially I'd rather be on the safe side and isolate Tahoe API throttling

**Comment by johnbaldwin** [comment]
Updated per PR change requests. Ran appsembler/api tests locally and passed. Will squash and merge

**Pull request merged**